### PR TITLE
executor: fix regular expression in json so that it could match identifer start with '$'

### LINKED ISSUE #29673: JSON_EXTRACT function does not get the expected result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
SELECT JSON_EXTRACT('{"properties": {"$type": "TiDB"}}','$.properties.$type')
```

### 2. What did you expect to see? (Required)

"TiDB"

### 3. What did you see instead (Required)

{"$type": "TiDB"}

### 4. What is your TiDB version? (Required)

5.7.25-TiDB-v5.1.0

### PR DESCRIPTION
**Pull request by znhh6018**
### What problem does this PR solve?

Issue Number: close #29673 

Problem Summary: 

### What is changed and how it works?
Modify json regular expressions to support syntax that start with '$' , e.g. '$.properties.$type' , we need to match "$type".

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: types/json/binary_test.go [hunk 1]
```diff
@@ -67,6 +70,10 @@ func TestBinaryJSONExtract(t *testing.T) {
 		// test extract with multi path expressions.
 		{bj1, []string{"$.a", "$[5]"}, mustParseBinaryFromString(t, `[[1, "2", {"aa": "bb"}, 4.0, {"aa": "cc"}]]`), true, nil},
 		{bj2, []string{"$.a", "$[0]"}, mustParseBinaryFromString(t, `[{"a": 1, "b": true}]`), true, nil},
+		{bj3, []string{`$.properties.$type`}, mustParseBinaryFromString(t, `"TiDB"`), true, nil},
```
Comment (Ehco1996) [inline types/json/binary_test.go:73]: L73~L76 is not the `multi path expressions.` please move this test to L69 and maybe add some `multi path` test case for `$`

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: types/json/binary_test.go (hunk 1 above)
Comment (znhh6018) [reply types/json/binary_test.go:70]: I've moved these tests and add multi path test case for $

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Review by dveeden — APPROVED** [review]
The basic functionality works as described:

With TiDB built from this PR:
```
sql> SELECT JSON_EXTRACT('{"$a": "b"}', '$.$a');
+-------------------------------------+
| JSON_EXTRACT('{"$a": "b"}', '$.$a') |
+-------------------------------------+
| "b"                                 |
+-------------------------------------+
1 row in set (0.0003 sec)
```

With MySQL 8.0.27:
```
sql> SELECT JSON_EXTRACT('{"$a": "b"}', '$.$a');
+-------------------------------------+
| JSON_EXTRACT('{"$a": "b"}', '$.$a') |
+-------------------------------------+
| "b"                                 |
+-------------------------------------+
1 row in set (0.0003 sec)
```

However there is some difference here:

With TiDB built from this PR:
```
sql> SELECT JSON_EXTRACT('{"$a$a": "b"}', '$.$a$a');
+-----------------------------------------+
| JSON_EXTRACT('{"$a$a": "b"}', '$.$a$a') |
+-----------------------------------------+
| NULL                                    |
+-----------------------------------------+
1 row in set (0.0014 sec)
```

And with MySQL 8.0.27:
```
sql> SELECT JSON_EXTRACT('{"$a$a": "b"}', '$.$a$a');
+-----------------------------------------+
| JSON_EXTRACT('{"$a$a": "b"}', '$.$a$a') |
+-----------------------------------------+
| "b"                                     |
+-----------------------------------------+
1 row in set (0.0009 sec)
```

**Pull request merged**